default/python/universe_generation/specials.py

```diff
@@ -1,76 +1,179 @@
-    # to add that special by testing its spawn rate
-    repeat_rate = {1: 0.08, 2: 0.05, 3: 0.01, 4: 0.00}
-    for univ_obj in universe_objects:
-        # for this universe object, find a suitable special
-        # start by shuffling our specials list, so each time the specials are considered in a new random order
-        random.shuffle(specials)
-        num_here = 0
-
-        # then, consider each special until one has been found or we run out of specials
-        # (the latter case means that no special is added to this universe object)
+        print("... {:30}: spawn rate {:2.3f} / spawn limit {}".
+              format(special, fo.special_spawn_rate(special), fo.special_spawn_limit(special)))
+
+    specials_timer.start("Filter By Base Chance")
+    objects_needing_specials = [obj for obj in universe_objects if random.random() < base_chance]
+
+    specials_timer.start("Compile num_placed statistics")
+    track_num_placed = {obj: 0 for obj in universe_objects}
+
+    print("Base chance for specials is {}. Placing specials on {} of {} ({:1.4f})objects"
+          .format(base_chance, len(objects_needing_specials), len(universe_objects),
+                  float(len(objects_needing_specials)) / len(universe_objects)))
+
+    specials_timer.start("Associate with Systems")
+    obj_tuple_needing_specials = set(zip(objects_needing_specials,
+                                         fo.objs_get_systems(objects_needing_specials),
+                                         calc_num_placed(objects_needing_specials)))
+
+    # Equal to the largest distance in WithinStarlaneJumps conditions
+    # GALAXY_DECOUPLING_DISTANCE is used as follows.  For any two or more objects
+    # at least GALAXY_DECOUPLING_DISTANCE appart you only need to check
+    # fo.special_locations once and then you can place as many specials as possible,
+    # subject to number restrictions.
+    #
+    # Organize the objects into sets where all objects are spaced GALAXY_DECOUPLING_DISTANCE
+    # appart.  Place a special on each one.  Repeat until you run out of specials or objects.
+    GALAXY_DECOUPLING_DISTANCE = 6
+
+    i_cache = 0
+    while obj_tuple_needing_specials:
+        systems_needing_specials = defaultdict(set)
+        for (obj, sys, n) in obj_tuple_needing_specials:
+            systems_needing_specials[sys].add((obj, sys, n))
+
+        print " Placing in " + str(len(systems_needing_specials)) + " locations remaining"
+
+        # Find a list of candidates all spaced GALAXY_DECORRECLATION_DISTANCE apart
+        specials_timer.start("Compose spaced list")
+        candidates = []
+        while systems_needing_specials:
+            specials_timer.start("Randos")
+            random_sys = random.choice(systems_needing_specials.values())
+            member = random.choice(list(random_sys))
+            candidates.append(member[0])
+            obj_tuple_needing_specials.remove(member)
+            if member[2] > 1:
```

Comment:
Unpack tuple to variables `objects_need_special, systems_for_object, number_to_place = memeber`